service/tablet_allocator.cc

```diff
@@ -727,6 +826,110 @@ class load_balancer {
+        // per-shard distribution as bad.
+        double shard_balance_threshold = div_ceil(total_load, total_shard_count);
+        // For determining impact on leaving, round down, because we don't want to consider movement which is within
+        // the best possible per-shard distribution as bad.
+        double leaving_shard_balance_threshold = total_load / total_shard_count;
+
+        auto new_shard_load_dst = nodes[dst.host].shards[dst.shard].tablet_count_per_table[table] + 1;
+        auto new_shard_load_src = nodes[src.host].shards[src.shard].tablet_count_per_table[table] - 1;
+
+        auto dst_shard_badness = (new_shard_load_dst - shard_balance_threshold) / total_load;
+
+        auto src_shard_badness = nodes[src.host].drained
+                ? 0 // Moving a tablet away from a drained node is always good.
+                : (leaving_shard_balance_threshold - new_shard_load_src) / total_load;
+
+        lblogger.trace("Table {} shard balance threshold: {}, src: {} ({:.4f}), dst: {} ({:.4f})", table,
+                       shard_balance_threshold, new_shard_load_src, src_shard_badness, new_shard_load_dst, dst_shard_badness);
+
+        // max number of tablets per node to keep perfect distribution.
+        double node_balance_threshold = div_ceil(total_load, node_count);
+        double leaving_node_balance_threshold = total_load / node_count;
+        auto dst_node_badness = (new_load_dst_node - node_balance_threshold) / total_load;
+
+        auto src_node_badness = nodes[src.host].drained
+                ? 0 // Moving a tablet away from a drained node is always good.
+                : (leaving_node_balance_threshold - new_load_src_node) / total_load;
+
+        if (src.host == dst.host) {
+            src_node_badness = 0;
+            dst_node_badness = 0;
+        }
+
+        lblogger.trace("Table {} node balance threshold: {}, src: {} ({:.4f}), dst: {} ({:.4f})", table,
+                       node_balance_threshold, new_load_src_node, src_node_badness, new_load_dst_node, dst_node_badness);
+
+        co_return migration_badness{
+            src_shard_badness,
+            src_node_badness,
+            dst_shard_badness,
+            dst_node_badness
+        };
+    }
+
+    future<migration_candidate> peek_candidate(node_load_map& nodes, shard_load& shard_info, tablet_replica src, tablet_replica dst) {
+        if (!_use_table_aware_balancing) {
+            co_return migration_candidate{peek_candidate(shard_info), src, dst, migration_badness{}};
+        }
+
+        std::map<migration_badness, table_id> best_tables;
+
+        for (auto&& [table, tablets] : shard_info.candidates) {
+            auto badness = co_await evaluate_candidate(nodes, table, src, dst);
+            best_tables[badness] = table;
+        }
+
+        if (best_tables.empty()) {
+            on_internal_error(lblogger, format("No candidates for migration on {}", src));
```

Comment:
This can only happen if `shard_info.candidates` is empty, so it's worth failing earlier

Reply:
Why is it worth failing earlier? 

Note that this validates precondition right before the place which expects best_tables to not be empty.

Reply:
> Why is it worth failing earlier?

From the code readability perspective. Currently it "converts" shard_info.candidates into best_tables, then checks that the latter is not empty and complains that "no candidates ..." otherwise. At this point it becomes clear, that the whole peek_candidate() wants non-empty list of such, but it could have been clear earlier

Reply:
I see, ok.